In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
df=pd.read_excel('/content/Healthcare dataset.xlsx')
df.head()

date   patient_id patient_gender  patient_age  patient_sat_score  \
0 2019-04-01  874-57-8352              M           38                NaN   
1 2019-04-01  259-51-9518              M           62                NaN   
2 2019-04-01  417-42-4910              F           79                NaN   
3 2019-04-01  428-40-4491              F           46                NaN   
4 2019-04-01  868-10-5414              F           55                NaN   

      Full Name        patient_race  patient_admin_flag  patient_waittime  \
0  V Matyushenko              White               False                50   
1    T Rolingson              Asian               False                12   
2       V Clelle  Two or More Races               False                45   
3   J Antonnikov   African American                True                45   
4       R Fulker   African American                True                24   

  department_referral Moment  
0                 NaN     AM  
1                 NaN     AM  
2                 NaN     PM  
3                 NaN     PM  
4                 NaN     PM

In [2]:
df.tail()

date   patient_id patient_gender  patient_age  patient_sat_score  \
9211 2020-10-30  628-46-3223              F           13                5.0   
9212 2020-10-30  278-89-6924              M           29                2.0   
9213 2020-10-30  618-76-9696              M            7               10.0   
9214 2020-10-30  454-91-6429              F           28                8.0   
9215 2020-10-30  734-88-0672              M           43                4.0   

      Full Name        patient_race  patient_admin_flag  patient_waittime  \
9211    W Wardle  Two or More Races                True                45   
9212    B Sandom              Asian                True                18   
9213   I Rummins  Two or More Races                True                15   
9214  B Hodinton  Two or More Races                True                26   
9215  R Clemenzi              Asian               False                51   

     department_referral Moment  
9211         Orthopedics     AM  
9212    General Practice     AM  
9213    General Practice     PM  
9214                 NaN     PM  
9215                 NaN     AM

In [3]:
df.shape

(9216, 11)

In [4]:
df.describe()

date  patient_age  patient_sat_score  \
count                           9216  9216.000000        2517.000000   
mean   2020-01-16 02:22:48.749999872    39.855143           4.992054   
min              2019-04-01 00:00:00     1.000000           0.000000   
25%              2019-08-24 00:00:00    20.000000           2.000000   
50%              2020-01-17 00:00:00    39.000000           5.000000   
75%              2020-06-09 00:00:00    60.000000           8.000000   
max              2020-10-30 00:00:00    79.000000          10.000000   
std                              NaN    22.755125           3.138043   

       patient_waittime  
count       9216.000000  
mean          35.259874  
min           10.000000  
25%           23.000000  
50%           35.000000  
75%           48.000000  
max           60.000000  
std           14.735323

In [5]:
df.info(0)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9216 entries, 0 to 9215
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   date                 9216 non-null   datetime64[ns]
 1   patient_id           9216 non-null   object        
 2   patient_gender       9216 non-null   object        
 3   patient_age          9216 non-null   int64         
 4   patient_sat_score    2517 non-null   float64       
 5   Full Name            9216 non-null   object        
 6   patient_race         9216 non-null   object        
 7   patient_admin_flag   9216 non-null   bool          
 8   patient_waittime     9216 non-null   int64         
 9   department_referral  3816 non-null   object        
 10  Moment               9216 non-null   object        
dtypes: bool(1), datetime64[ns](1), float64(1), int64(2), object(6)
memory usage: 729.1+ KB


In [6]:
# Mark missing
df['score_was_missing'] = df['patient_sat_score'].isna().astype(int)# check the patient_sat column and look if value is null or not then assign 1 if null, 0 if not null

# Fill missing safely
df['patient_sat_score'] = df['patient_sat_score'].fillna(
    df['patient_sat_score'].median()
)


df['patient_sat_score'].median()
→ finds the middle value of all existing scores
(ignores missing ones)

.fillna(median)
→ replaces only missing values with that median
→ existing values stay unchanged

.astype(int)
→ converts:
True → 1
False → 0

In [7]:
df['patient_sat_score'].isna().sum()#check if still missing?


np.int64(0)

In [8]:
df.head()

date   patient_id patient_gender  patient_age  patient_sat_score  \
0 2019-04-01  874-57-8352              M           38                5.0   
1 2019-04-01  259-51-9518              M           62                5.0   
2 2019-04-01  417-42-4910              F           79                5.0   
3 2019-04-01  428-40-4491              F           46                5.0   
4 2019-04-01  868-10-5414              F           55                5.0   

      Full Name        patient_race  patient_admin_flag  patient_waittime  \
0  V Matyushenko              White               False                50   
1    T Rolingson              Asian               False                12   
2       V Clelle  Two or More Races               False                45   
3   J Antonnikov   African American                True                45   
4       R Fulker   African American                True                24   

  department_referral Moment  score_was_missing  
0                 NaN     AM                  1  
1                 NaN     AM                  1  
2                 NaN     PM                  1  
3                 NaN     PM                  1  
4                 NaN     PM                  1

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9216 entries, 0 to 9215
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   date                 9216 non-null   datetime64[ns]
 1   patient_id           9216 non-null   object        
 2   patient_gender       9216 non-null   object        
 3   patient_age          9216 non-null   int64         
 4   patient_sat_score    9216 non-null   float64       
 5   Full Name            9216 non-null   object        
 6   patient_race         9216 non-null   object        
 7   patient_admin_flag   9216 non-null   bool          
 8   patient_waittime     9216 non-null   int64         
 9   department_referral  3816 non-null   object        
 10  Moment               9216 non-null   object        
 11  score_was_missing    9216 non-null   int64         
dtypes: bool(1), datetime64[ns](1), float64(1), int64(3), object(6)
memory usage: 801.1+ KB


In [10]:
df['department_referral'] = df['department_referral'].fillna('No Referral')
df['department_referral'].isna().sum()#check if still null?


np.int64(0)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9216 entries, 0 to 9215
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   date                 9216 non-null   datetime64[ns]
 1   patient_id           9216 non-null   object        
 2   patient_gender       9216 non-null   object        
 3   patient_age          9216 non-null   int64         
 4   patient_sat_score    9216 non-null   float64       
 5   Full Name            9216 non-null   object        
 6   patient_race         9216 non-null   object        
 7   patient_admin_flag   9216 non-null   bool          
 8   patient_waittime     9216 non-null   int64         
 9   department_referral  9216 non-null   object        
 10  Moment               9216 non-null   object        
 11  score_was_missing    9216 non-null   int64         
dtypes: bool(1), datetime64[ns](1), float64(1), int64(3), object(6)
memory usage: 801.1+ KB


In [12]:
df.isna().sum().sum()


np.int64(0)

In [13]:
# How long do patients wait?
df['patient_waittime'].describe()

count    9216.000000
mean       35.259874
std        14.735323
min        10.000000
25%        23.000000
50%        35.000000
75%        48.000000
max        60.000000
Name: patient_waittime, dtype: float64

count = 9216
👉You have data for 9,216 patients

Average waiting time
mean = 35.26 minutes
👉On average, a patient waits about 35 minutes

Middle waiting time
50% (median) = 35 minutes
👉 Half of patients waited less than 35 min
👉 Half waited more than 35 min
(This confirms the average is reasonable)

Shortest wait
min = 10 minutes
👉 Fastest patient waited 10 minutes

Longest wait
max = 60 minutes
👉 Slowest patient waited 1 hour

Most common waiting range (VERY IMPORTANT)
25% = 23 minutes
→ 25% patients waited less than 23 min

75% = 48 minutes
→ 75% patients waited less than 48 min
👉 So most patients wait between 23 and 48 minutes

In [14]:
#do patients who wait longer feel less satisfied?
df.groupby('patient_sat_score')['patient_waittime'].mean()


patient_sat_score
0.0     36.108108
1.0     35.373984
2.0     36.887255
3.0     35.092105
4.0     34.717742
5.0     35.221243
6.0     35.588745
7.0     35.910156
8.0     33.655963
9.0     35.265766
10.0    35.248869
Name: patient_waittime, dtype: float64

Average waiting time remains similar across all satisfaction levels, indicating wait time alone may not drive patient satisfaction.

In [17]:
#Check AM vs PM waiting time
df.groupby('Moment')['patient_waittime'].mean()

Moment
AM    35.533679
PM    34.983202
Name: patient_waittime, dtype: float64

Average patient waiting time is nearly identical in morning and evening hours.

In [18]:
#Referral vs No Referral waiting time
df.groupby('department_referral')['patient_waittime'].mean()



department_referral
Cardiology          35.354839
Gastroenterology    35.831461
General Practice    34.912500
Neurology           36.803109
No Referral         35.293148
Orthopedics         34.982915
Physiotherapy       36.565217
Renal               34.697674
Name: patient_waittime, dtype: float64

Average waiting time is similar across referral types, with slightly higher waits for specialized departments like Neurology

In [19]:
#Who skips satisfaction feedback?
df.groupby('score_was_missing')['patient_waittime'].mean()


score_was_missing
0    35.351212
1    35.225556
Name: patient_waittime, dtype: float64

1->patients who did NOT give feedback  0-> patients who gave feedback

Patients who skipped the satisfaction feedback had almost the same waiting time as those who gave feedback.


In [23]:
#This creates groups like young, adult, middle-aged, senior.
df['age_group'] = pd.cut(df['patient_age'],
    bins=[0, 18, 35, 50, 65, 100],
    labels=['0-18', '19-35', '36-50', '51-65', '65+']
)

In [24]:
#Compare waiting time by age group
df.groupby('age_group')['patient_waittime'].mean()


/tmp/ipython-input-1075629247.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('age_group')['patient_waittime'].mean()


age_group
0-18     35.318483
19-35    35.499748
36-50    35.351914
51-65    34.864583
65+      35.210266
Name: patient_waittime, dtype: float64

Average patient waiting time remains consistent across different age groups.

In [25]:
df.groupby('department_referral')['patient_sat_score'].mean()


department_referral
Cardiology          5.040323
Gastroenterology    5.241573
General Practice    5.016848
Neurology           5.077720
No Referral         4.987593
Orthopedics         4.958794
Physiotherapy       4.996377
Renal               4.883721
Name: patient_sat_score, dtype: float64

Gastroenterology shows the highest patient satisfaction, followed by Neurology and Cardiology, while satisfaction levels remain fairly similar across all departments.